### PR TITLE
Add a new service ThuderService for Oracle Cloud integration

### PR DESCRIPTION
**Pull request by anbrenna**

### CODE REVIEW

File: cmd/serve.go [hunk 1]
```diff
@@ -372,3 +382,54 @@ func serveRepoService(config config.Configuration, wg *sync.WaitGroup) {
...
+	service := &thunderservice_service.ThunderService{
+		Config:   config,
+		Database: db,
+		TimeNow:  time.Now,
+		Log:      log,
+	}
+	service.Init()
+
+	/*
+		auth := apiservice_auth.BuildAuthenticationProvider(config.APIService.AuthenticationProvider, time.Now, log)
```
Comment (cirelli94) [inline cmd/serve.go:405]: I think you forgot this comments inside the code!

File: cmd/serve.go [hunk 2]
```diff
@@ -372,3 +382,54 @@ func serveRepoService(config config.Configuration, wg *sync.WaitGroup) {
...
+		auth.Init()
+	*/
+	router := mux.NewRouter()
+	ctrl := &thunderservice_controller.ThunderController{
+		Config:  config,
+		Service: service,
+		TimeNow: time.Now,
+		Log:     log,
+		//		Authenticator: auth,
```
Comment (cirelli94) [inline cmd/serve.go:414]: ...and this comment too!

File: cmd/serve.go [hunk 3]
```diff
@@ -372,3 +382,54 @@ func serveRepoService(config config.Configuration, wg *sync.WaitGroup) {
...
+	}
+	thunderservice_controller.SetupRoutesForThunderController(router, ctrl)
+
+	var logRouter http.Handler
+	if config.ChartService.LogHTTPRequest {
```
Comment (cirelli94) [inline cmd/serve.go:419]: You should use the ThunderService configuration!?

File: config/config.go [hunk 4]
```diff
@@ -154,6 +156,20 @@ type ChartService struct {
...
+	// RemoteEndpoint contains the endpoint used to connect to the ThunderService
+	RemoteEndpoint string
+	// BindIP contains the bind ip
+	BindIP string
+	// Port contains the port of the internal http server
+	Port uint16
+	// LogHTTPRequest enable the logging of the internal http serverl
+	LogHTTPRequest bool
+	// LogHTTPRequest enable the logging of the received messages
+	LogMessages bool
```
Comment (cirelli94) [inline config/config.go:170]: Is this needed?

File: go.mod [hunk 5]
```diff
@@ -29,6 +29,8 @@ require (
 	github.com/leandro-lugaresi/hub v1.1.1
 	github.com/lucasb-eyer/go-colorful v1.2.0
 	github.com/niemeyer/pretty v0.0.0-20200227124842-a10e7caefd8e // indirect
+	github.com/oracle/oci-go-sdk v24.3.0+incompatible // indirect
```
Comment (cirelli94) [inline go.mod:32]: This "incompatible" dependency looks strange to me!
Try to clean your go.mod with `go mod tidy` to check it's needed!

File: thunder-service/controller/recommendations.go [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+	"errors"
+	"net/http"
+
+	"github.com/ercole-io/ercole/v2/utils"
+)
+
+//GetOCRecommendations get recommendation from Oracle Cloud
+func (ctrl *ThunderController) GetOCRecommendations(w http.ResponseWriter, r *http.Request) {
+
+	// da recuprare come parametro dalla request
```
Comment (cirelli94) [inline thunder-service/controller/recommendations.go:29]: Is this comment needed?

File: thunder-service/controller/controller.go [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+
+	"github.com/ercole-io/ercole/v2/config"
+	"github.com/ercole-io/ercole/v2/logger"
+	"github.com/ercole-io/ercole/v2/thunder-service/service"
+)
+
+// ThunderControllerInterface is a interface that wrap methods used to inserting events in the queue
+type ThunderControllerInterface interface {
+
+	// GetOCRecommendations get recommendations from Oracle Cloud Infracstructure
```
Comment (cirelli94) [inline thunder-service/controller/controller.go:31]: Infra**c**structure typo

File: thunder-service/controller/controller.go [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+	GetOCRecommendations(w http.ResponseWriter, r *http.Request)
+
+	// GetOCRecommendations get recommendations from Oracle Cloud Infracstructure
+	GetOCRecommendationsWithCategory(w http.ResponseWriter, r *http.Request)
+
+	//AuthenticateMiddleware(h http.Handler) http.Handler
```
Comment (cirelli94) [inline thunder-service/controller/controller.go:37]: Is this needed?

File: thunder-service/controller/controller.go [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+	"github.com/ercole-io/ercole/v2/thunder-service/service"
+)
+
+// ThunderControllerInterface is a interface that wrap methods used to inserting events in the queue
+type ThunderControllerInterface interface {
+
+	// GetOCRecommendations get recommendations from Oracle Cloud Infracstructure
+	GetOCRecommendations(w http.ResponseWriter, r *http.Request)
+
+	// GetOCRecommendations get recommendations from Oracle Cloud Infracstructure
```
Comment (cirelli94) [inline thunder-service/controller/controller.go:34]: Infracstructure typo

File: thunder-service/controller/controller.go [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+	GetOCRecommendationsWithCategory(w http.ResponseWriter, r *http.Request)
+
+	//AuthenticateMiddleware(h http.Handler) http.Handler
+}
+
+// ThunderController is the struct used to handle the requests from agents and contains the concrete implementation of AlertQueueControllerInterface
```
Comment (cirelli94) [inline thunder-service/controller/controller.go:40]: AlertQueueControllerInterface ?! :)

File: thunder-service/controller/recommendations.go [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+	compartmentId := "ocid1.tenancy.oc1..aaaaaaaazizzbqqbjv2se3y3fvm5osfumnorh32nznanirqoju3uks4buh4q"
```
Comment (cirelli94) [inline thunder-service/controller/recommendations.go:30]: This parameter shouldn't be hardcoded in the project!
We should retrieve it from somewhere...

File: thunder-service/controller/recommendations.go [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+		return
+	}
+
+	utils.WriteJSONResponse(w, http.StatusOK, data)
+}
+
+//GetOCRecommendations get recommendation with Category from Oracle Cloud
+func (ctrl *ThunderController) GetOCRecommendationsWithCategory(w http.ResponseWriter, r *http.Request) {
+
+	// compartmentId parameter has to be retrieved from configuration
```
Comment (cirelli94) [inline thunder-service/controller/recommendations.go:47]: The same as above: we should retrieve the compartmentId from somewhere, maybe from a db collection?

File: thunder-service/controller/routing.go [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package controller
+
+import (
+	"net/http"
+
+	//	"github.com/goji/httpauth"
```
Comment (cirelli94) [inline thunder-service/controller/routing.go:21]: Remove the comment!!!

File: thunder-service/controller/routing.go [hunk 14]
```diff
@@ -0,0 +1,40 @@
...
+)
+
+// SetupRoutesForThunderController setup the routes of the router using the handler in the controller as http handler
+func SetupRoutesForThunderController(router *mux.Router, ctrl ThunderControllerInterface) {
+	router.HandleFunc("/ping", func(w http.ResponseWriter, r *http.Request) {
+		w.Write([]byte("ABPong"))
+	})
+
+	router = router.NewRoute().Subrouter()
+	//	router.Use(ctrl.AuthenticateMiddleware())
```
Comment (cirelli94) [inline thunder-service/controller/routing.go:32]: Comment!

File: thunder-service/controller/routing.go [hunk 15]
```diff
@@ -0,0 +1,40 @@
...
+	"net/http"
+
+	//	"github.com/goji/httpauth"
+	"github.com/gorilla/mux"
+)
+
+// SetupRoutesForThunderController setup the routes of the router using the handler in the controller as http handler
+func SetupRoutesForThunderController(router *mux.Router, ctrl ThunderControllerInterface) {
+	router.HandleFunc("/ping", func(w http.ResponseWriter, r *http.Request) {
+		w.Write([]byte("ABPong"))
```
Comment (cirelli94) [inline thunder-service/controller/routing.go:28]: Keep it as Pong :)

File: thunder-service/database/database.go [hunk 16]
```diff
@@ -0,0 +1,79 @@
...
+	"github.com/ercole-io/ercole/v2/logger"
+	"go.mongodb.org/mongo-driver/mongo"
+	"go.mongodb.org/mongo-driver/mongo/options"
+)
+
+// MongoDatabaseInterface is a interface that wrap methods used to perform CRUD operations in the mongodb database
+type MongoDatabaseInterface interface {
+	// Init initializes the connection to the database
+	Init()
+	// FindHostData find a host data
```
Comment (cirelli94) [inline thunder-service/database/database.go:33]: Clear the comments!!!!!!!1

File: thunder-service/service/recommendations.go [hunk 17]
```diff
@@ -0,0 +1,180 @@
...
+	"fmt"
+	"strconv"
+
+	"github.com/ercole-io/ercole/v2/model"
+	"github.com/oracle/oci-go-sdk/v45/common"
+	"github.com/oracle/oci-go-sdk/v45/example/helpers"
+	"github.com/oracle/oci-go-sdk/v45/optimizer"
+)
+
+// 0 high-availability-name high-availability-desc 0 ocid1.optimizercategory.oc1..aaaaaaaa34bnli7iarbwzpaz6abuup2egv2hcdyiixzellvxplhmceqvzula
```
Comment (cirelli94) [inline thunder-service/service/recommendations.go:30]: Commentsssssssssss

File: thunder-service/service/service.go [hunk 18]
```diff
@@ -0,0 +1,76 @@
+// Copyright (c) 2020 Sorint.lab S.p.A.
```
Comment (cirelli94) [inline thunder-service/service/service.go:1]: a lot of comments here too......

**Review by cirelli94 — CHANGES_REQUESTED** [review]

File: thunder-service/service/recommendations.go [hunk 19]
```diff
@@ -0,0 +1,164 @@
...
+
+	return GetOCListRecommendationsWithCategory(compartmentId)
+}
+
+func GetOCListRecommendations(compartmentId string) ([]model.Recommendation, error) {
+	// Create a default authentication provider that uses the DEFAULT
+	// profile in the configuration file.
+	// Refer to <see href="https://docs.cloud.oracle.com/en-us/iaas/Content/API/Concepts/sdkconfig.htm#SDK_and_CLI_Configuration_File>the public documentation</see> on how to prepare a configuration file.
+	client, err := optimizer.NewOptimizerClientWithConfigurationProvider(common.DefaultConfigProvider())
+	helpers.FatalIfError(err)
```
Comment (cirelli94) [inline thunder-service/service/recommendations.go:45]: Do not fatal if error, but return an error value!!!!!!!

**Review by cirelli94 — CHANGES_REQUESTED** [review]

File: thunder-service/service/recommendations.go (hunk 19 above)
Comment (anbrenna) [reply thunder-service/service/recommendations.go:45]: Updated with the correct management of the error

**Review by cirelli94 — CHANGES_REQUESTED** [review]

**Pull request merged**